downstairs/src/lib.rs

```diff
@@ -989,101 +989,61 @@ where
     //  │ recv_task  │       │   reply_task   │
     //  └─────┬──────┘       └────────▲───────┘
     //        │                       │
-    //        │message          reply │responses,
-    //        │channel         channel│invalid frame
+    //        │downstairs       reply │responses,
+    //        │handle          channel│invalid frame
     //        │                       │errors, pings
     //        │                       │
-    //        │              ┌────────┴───────┐  ┌──────┐ monitors other tasks
-    //        └──────────────►   proc_task    │  │ proc │ and aborts them if
-    //                       └──┬───────────▲─┘  └──────┘ any have stopped
-    //                          │           │
-    //                       add│work    new│work
-    //   per-connection         │           │
-    //  ======================= │ ========= │ =======
-    //   shared state        ┌──▼───────────┴─┐
-    //                       │   Downstairs   │
-    //                       └────────────────┘
-
-    // We rely on backpressure in the upstairs to limit total jobs in flight, so
-    // these channels can be unbounded.
-    let (msg_channel_tx, msg_channel_rx) = mpsc::unbounded_channel();
-    let mut recv_task = tokio::spawn(recv_task(
-        fr,
-        msg_channel_tx,
-        log.new(o!("task" => "recv")),
-    ));
-
+    //        │                       │
+    //        └─────────────────┐     │
+    //                          │     │
+    //                          │     │
+    //                          │     │
+    //   per-connection state   │     │
+    //  ======================= │ === │ ===============================
+    //   common state      ┌────▼─────┴──────┐
+    //                     │ Downstairs::run │
+    //                     └─────────────────┘
+    //
+    // Task lifetimes are handled automatically; there are a variety of possible
+    // orderings.
+    //
+    // ## Error in `recv_task`
+    // - `recv_task` will proactively send `Message::ConnectionClosed` to the
+    //   `Downstairs`, then exit
+    // - The `Downstairs` will drop the `ConnectionState`, causing `reply_task`
+    //   to exit because its `mpsc::Receiver` will read `None`
+    //
+    // ## Intentional disconnection by the `Downstairs`
+    // - The `Downstairs` drops the `ConnectionState
+    // - `reply_task` stops as above when its `Receiver` reads `None`
+    // - The `ConnectionState` owns a `DropGuard` and `recv_task` polls the
+    //   corresponding `CancellationToken`; when dropped, `recv_task` exits.
+    //
+    // ## Error in `reply_task`
+    // - `reply_task` will exit upon error
+    // - When the `Downstairs` tries to write to it, the write will fail
```

Comment:
Should we be more proactive here, e.g. stopping `recv_task` if `reply_task` exits?

Reply:
We should stop trying to do any work once we know the connection is closed.
If reply_task ends, will that signal to `Downstairs::run` that it should stop as well?
That's not really an answer to your question directly, but maybe if recv task ends,
it won't be trying to put more work on the queue.

Reply:
I made `recv_task` responsible for monitoring `reply_task` as well in aa6167a69b57, so it should cleanly shut down under all circumstances.